### ISSUE TITLE
[Clang] Are `{ std::is_integral_v<T> } -> std::same_as<bool>` and `{ std::is_integral_v<T> }` different constraints?

### ISSUE THREAD

**Issue by haoyang9804** [llvm/llvm-project#62328]
[Godbolt code](https://cute.godbolt.org/z/sMvY13jGf)
Code:
```C++
#include <type_traits>
#include <iostream>

template <typename T>  
concept Basic = requires(T t) {  
    { std::is_integral_v<T> } -> std::same_as<bool>;  
};  

template <typename T>  
concept Basic2 = requires(T t) {  
    { std::is_integral_v<T> };
};  

template <typename T>  
concept Basic3 = std::is_integral_v<T>;  

int main() {
    std::cout << Basic<int> << std::endl; // why this output 0??
    std::cout << Basic2<int> << std::endl; // should output 1
    std::cout << Basic3<int> << std::endl; // should output 1
}
```
I think `->std::same_as<bool>` is used to enforce that the result of an expression should be of type bool. And obviously, `std::is_integral_v<T> ` returns a bool value. Why will the addition of this return type restriction cause different output?

**Comment by chrchr-github** [comment]
MSVC and GCC can even tell the reason: https://cute.godbolt.org/z/shqbd5nrW

**Comment by shafik** [comment]
New to me but this is due to [expr.prim.req.compund p1.3](https://eel.is/c++draft/expr.prim.req.compound#1.3)

> If the return-type-requirement is present, then:
> - (1.3.1) Substitution of template arguments (if any) into the return-type-requirement is performed.
> - (1.3.2) The immediately-declared constraint ([temp.param]) of the type-constraint for decltype((E)) shall be satisfied.


and it has the following example

```cpp
[Example 1: Given concepts C and D,

requires {
  { E1 } -> C;
  { E2 } -> D<A1, ⋯, An>;
};

is equivalent to

requires {
  E1; requires C<decltype((E1))>;
  E2; requires D<decltype((E2)), A1, ⋯, An>;
};

(including in the case where n is zero). — end example]
```

The result of `decltype((std::is_integral_v<int>))` will be `const bool&`

**Comment by shafik** [comment]
Worth noting, that this is b/c it is an `lvalue` if you had something like this:

```cpp
template <typename T>
bool f() {
    return std::is_integral_v<T>;
}

template <typename T>  
concept A = requires(T t) {  
    { f<T>() } -> std::same_as<bool>;  
};  
```

You would obtain the result you expect b/c `f<T>()` produces a `prvalue`

**Comment by shafik** [comment]
CC @zygoloid this result feels surprising, do you have any insights on how to think about this better?

**Comment by zygoloid** [comment]
> MSVC and GCC can even tell the reason: https://cute.godbolt.org/z/shqbd5nrW

As does Clang, for what it's worth:

> [...] because 'std::is_same_v<const _Bool &, _Bool>' evaluated to false

We could special-case uses of the `std::same_as` concept and give a custom diagnostic; that would probably help here.

---

> CC @zygoloid this result feels surprising, do you have any insights on how to think about this better?

I think there are two surprising things here:

1) compound requirements in `requires` expressions [use the non-*id-expression* `decltype` rules](https://eel.is/c++draft/expr.prim.req#compound-1.3.2.sentence-1) to pick a type, rather than constraining the (non-reference) type of the expression
2) `std::is_integral_v<T>` is a variable template, so is an lvalue, so `decltype` applied to it produces a reference type

We can simplify this a little and still get a surprise:

```
template<typename T> concept thing = requires (T t) {
    {t} -> std::same_as<T>;
};
```

Here, `thing<int>` is not satisfied, but `thing<int&>` is, because the implied `decltype((t))` adds a `&`. Usually, `decltype`'s reference-adding behavior is not so surprising, because you wrote `decltype` and so are expecting it to do its thing.

**Comment by shafik** [comment]
> > CC @zygoloid this result feels surprising, do you have any insights on how to think about this better?
> 
> I think there are two surprising things here:
> 
> 1. compound requirements in `requires` expressions [use the non-_id-expression_ `decltype` rules](https://eel.is/c++draft/expr.prim.req#compound-1.3.2.sentence-1) to pick a type, rather than constraining the (non-reference) type of the expression
> 2. `std::is_integral_v<T>` is a variable template, so is an lvalue, so `decltype` applied to it produces a reference type
> 
> We can simplify this a little and still get a surprise:
> 
> ```
> template<typename T> concept thing = requires (T t) {
>     {t} -> std::same_as<T>;
> };
> ```
> 
> Here, `thing<int>` is not satisfied, but `thing<int&>` is, because the implied `decltype((t))` adds a `&`. Usually, `decltype`'s reference-adding behavior is not so surprising, because you wrote `decltype` and so are expecting it to do its thing.

Yeah, the simpler example does indeed feel worse and yes when you explicitly use `decltype` these behavior are more obvious.

I am wondering why it was specified this way, would more intuitive behavior break other uses?

**Comment by zygoloid** [comment]
It is -- sometimes -- useful to be able to constrain both the type and value category of an expression at the same time. For example:

```
template<typename T>
concept iterator = requires (T it) { {*it} -> same_as<typename T::value_type&>; };
```

Perhaps the situation would be better if we had a standard library concept like `same_as` for "same as `T`, except for references and qualifications, and convertible to `T`". The lack of such a thing spills over into the standard library specification using `same_as` where it's not really appropriate to be that strict. For example, if `vector::max_size()` is declared to return `const std::size_t&` instead of `std::size_t`, then `vector` [stops being a *reservable-container*](https://eel.is/c++draft/ranges#range.utility.conv.general-3), which seems like a (minor) mistake in the design of the concept to me.

**Comment by zygoloid** [comment]
Back to the case in hand: it looks like Clang's behavior is correct, but we do potentially have the opportunity to improve our diagnostics. I think there are two things we could do:

1) Special-case the diagnostics for failed satisfaction checks for `std::same_as<T, U>`, and report them as "because `T` (aka ...) is not the same type as `U` (aka ...)", rather than descending into the implementation details of libstdc++ or libc++ and not producing any insight beyond `is_same_v<T, U>` evaluating to false.
2) Special-case the situation where we have `{expr} -> SomeConcept`, where the type of `expr` satisfies `SomeConcept` but `decltype((expr))` does not, and produce a diagnostic explaining that the type being constrained is a reference type.

I think (1) is worth doing. I'm on the fence about (2), but I'm nervous because merely checking whether the expression's type satisfies `SomeConcept` can trigger non-immediate-context errors -- we at least can't do (2) unless we know we're going to reject the program, and will need to be careful to swallow diagnostics from the attempted concept evaluation. (But what if it leads to a fatal error, such as hitting the template instantiation depth?) So I'm inclined to say we should not do (2) for now.

Separately, the diagnostic we're currently producing is a little broken; it looks like we're using the wrong `LangOptions` when converting the type to a string:

> [...] because 'std::is_same_v<const _Bool &, _Bool>' evaluated to false

Note the `_Bool` here which should be `bool`. We should track that down and fix it.

**Comment by haoyang9804** [comment]
> New to me but this is due to [expr.prim.req.compund p1.3](https://eel.is/c++draft/expr.prim.req.compound#1.3)
> 
> > If the return-type-requirement is present, then:
> > 
> > * (1.3.1) Substitution of template arguments (if any) into the return-type-requirement is performed.
> > * (1.3.2) The immediately-declared constraint ([temp.param]) of the type-constraint for decltype((E)) shall be satisfied.
> 
> and it has the following example
> 
> ```c++
> [Example 1: Given concepts C and D,
> 
> requires {
>   { E1 } -> C;
>   { E2 } -> D<A1, ⋯, An>;
> };
> 
> is equivalent to
> 
> requires {
>   E1; requires C<decltype((E1))>;
>   E2; requires D<decltype((E2)), A1, ⋯, An>;
> };
> 
> (including in the case where n is zero). — end example]
> ```
> 
> The result of `decltype((std::is_integral_v<int>))` will be `const bool&`

Thanks for ur reply. I have modified my code based on this rule in [here](https://cute.godbolt.org/z/hPa6cs769).

```C++
#include <type_traits>
#include <iostream>

template <typename T>  
concept Basic = requires(T t) {  
    { std::is_integral_v<T> } -> std::same_as<const bool&>;  
};  

template <typename T>  
concept Basic2 = requires(T t) {  
    { std::is_integral_v<T> };
};

template <typename T>  
concept Basic3 = std::is_integral_v<T>;

template <typename T>  
concept Basic4 = requires(T t) {  
    std::is_integral_v<T>;
    requires std::same_as<decltype(std::is_integral_v<T>), const bool&>;  
};



int main() {
    std::cout << Basic<int> << std::endl;  // output 1
    std::cout << Basic2<int> << std::endl; // output 1
    std::cout << Basic3<int> << std::endl; // output 1
    std::cout << Basic4<int> << std::endl; // output 0
}
```
According to your feedback, the first and the fourth outputs should both be 1. But the fourth is not.

**Comment by zygoloid** [comment]
```
    requires std::same_as<decltype(std::is_integral_v<T>), const bool&>;
```
You missed one of the two pairs of parentheses around the operand of `decltype` here.

**Comment by haoyang9804** [comment]
> ```
>     requires std::same_as<decltype(std::is_integral_v<T>), const bool&>;
> ```
> 
> You missed one of the two pairs of parentheses around the operand of `decltype` here.

Wow. Thanks.
